server/src/MessageHandler.ts

```diff
@@ -5,77 +5,204 @@ import {
   Range,
   Location,
   DeclarationParams,
+  CompletionParams,
+  CompletionList,
+  CompletionItem,
+  Position,
 } from 'vscode-languageserver'
 import * as util from './util'
 import { fullDocumentRange } from './provider'
-import { getDMMF } from '@prisma/sdk'
-import { TextDocument, Position } from 'vscode-languageserver-textdocument'
+import { TextDocument } from 'vscode-languageserver-textdocument'
 import format from './format'
+import {
+  getSuggestionsForAttributes,
+  getSuggestionsForTypes,
+  getSuggestionForBlockTypes,
+  getSuggestionForFirstInsideBlock,
+  getSuggestionForSupportedFields,
+  getSuggestionsForInsideAttributes,
+} from './completions'
 
-function getWordAtPosition(document: TextDocument, position: Position): string {
-  const currentLine = document.getText({
-    start: { line: position.line, character: 0 },
-    end: { line: position.line, character: 9999 },
+function getCurrentLine(document: TextDocument, line: number) {
+  return document.getText({
+    start: { line: line, character: 0 },
+    end: { line: line, character: 9999 },
   })
+}
+
+function convertDocumentTextToTrimmedLineArray(
+  document: TextDocument,
+): Array<string> {
+  return Array(document.lineCount)
+    .fill(0)
+    .map((_, i) => getCurrentLine(document, i).trim())
+}
+
+function isFirstInsideBlock(position: Position, currentLine: string): boolean {
+  if (currentLine.trim().length === 0) {
+    return true
+  }
+
+  const stringTillPosition = currentLine.slice(0, position.character)
+  const matchArray = /\w+/.exec(stringTillPosition)
+
+  if (!matchArray) {
+    return true
+  }
+  return (
+    matchArray.length === 1 &&
+    matchArray.index !== undefined &&
+    stringTillPosition.length - matchArray.index - matchArray[0].length === 0
+  )
+}
+
+export function getWordAtPosition(
+  document: TextDocument,
+  position: Position,
+): string {
+  const currentLine = getCurrentLine(document, position.line)
+  const stringTillPosition = currentLine.slice(0, position.character)
+  if (stringTillPosition.endsWith('@@')) {
+    return '@@'
+  }
+  if (stringTillPosition.endsWith('@')) {
+    return '@'
+  }
   // search for the word's beginning and end
-  const beginning = currentLine.slice(0, position.character + 1).search(/\S+$/)
-  const end = currentLine.slice(position.character).search(/\W/)
+  const beginning: number = currentLine
+    .slice(0, position.character + 1)
+    .search(/\S+$/)
+  const end: number = currentLine.slice(position.character).search(/\W/)
   if (end < 0) {
     return ''
   }
   return currentLine.slice(beginning, end + position.character)
 }
 
-export async function handleDefinitionRequest(
+export class MyBlock {
```

Comment:
Is the name `MyBlock` intentional? What does this represent?

Reply:
When using the prismafile parser, there already as an object Block, but since I can't use this anymore, I called this MyBlock to not get confused with the other Object. Since I am not using the other dependency anmore, I guess I'll just rename this to `Block` again. This data structure is used to store all Blocks in a schema so I don't have parse every line in a string multiple times.